src/common/proxy_wasm/ngx_proxy_wasm_foreign_call.c

```diff
@@ -0,0 +1,293 @@
+    size_t size)
+{
+    ngx_chain_t              *cl;
+    ngx_proxy_wasm_exec_t    *pwexec = call->pwexec;
+    ngx_http_wasm_req_ctx_t  *rctx = call->rctx;
+
+    ngx_wa_assert(pwexec);
+    ngx_wa_assert(size);
+
+    /* TODO: if size > NGX_MAX_ERROR_STR, split allocation into N buffers */
+
+    ngx_wa_assert(size <= NGX_MAX_ERROR_STR);
+
+    cl = ngx_wasm_chain_get_free_buf(pwexec->pool,
+                                     &rctx->free_bufs,
+                                     size, buf_tag,
+                                     1);
+    if (cl == NULL) {
+        return NGX_ERROR;
+    }
+
+    call->args_out = cl;
+
+    return NGX_OK;
+}
+
+
+static void
+ngx_proxy_wasm_foreign_call_buffer_write(ngx_proxy_wasm_foreign_call_t *call,
+    ngx_str_t *data)
+{
+    size_t      b_size;
+    ngx_buf_t  *b = call->args_out->buf;
+
+    b_size = b->end - b->start;
+
+    /* TODO: data->len > b_size */
+
+    ngx_wa_assert(data->len <= b_size);
+
+    if (data->len <= b_size) {
+        b->last = ngx_cpymem(b->last, data->data, data->len);
+    }
+}
+
+
+static void
+ngx_proxy_wasm_hfuncs_resolve_lua_handler(ngx_resolver_ctx_t *rslv_ctx)
+{
+#if (NGX_HAVE_INET6)
+    struct sockaddr_in6            *sin6;
+#endif
+    struct sockaddr_in             *sin;
+    u_char                         *p;
+    ngx_str_t                       args;
+    ngx_wasm_lua_ctx_t             *lctx;
+    ngx_wasm_socket_tcp_t          *sock;
+    ngx_proxy_wasm_foreign_call_t  *call;
+    u_short                         sa_family = AF_INET;
+    u_char                          buf[rslv_ctx->name.len +
+#if (NGX_HAVE_INET6)
+                                        sizeof(struct in6_addr) + 1];
+#else
+                                        sizeof(struct in_addr) + 1];
+#endif
+
+    if (rslv_ctx->addr.socklen == sizeof(struct sockaddr_in6)) {
+        sa_family = AF_INET6;
+    }
+
+    sock = (ngx_wasm_socket_tcp_t *) rslv_ctx->data;
+    lctx = sock->lctx;
+    call = (ngx_proxy_wasm_foreign_call_t *) sock->data;
+    p = buf;
+
+    ngx_memzero(buf, sizeof(buf));
+
+    if (rslv_ctx->state || !rslv_ctx->naddrs) {
+        p++;  /* 1st byte is address length; 0 if address not found */
+        goto not_found;
+    }
+
+    ngx_wa_assert(rslv_ctx->naddrs == 1);
+
+    switch (sa_family) {
+#if (NGX_HAVE_INET6)
+    case AF_INET6:
+          sin6 = (struct sockaddr_in6 *) rslv_ctx->addr.sockaddr;
+          *(p++) = sizeof(struct in6_addr);
+          p = ngx_cpymem(p, &sin6->sin6_addr, sizeof(struct in6_addr));
+          break;
+#endif
+    default: /* AF_INET */
+          sin = (struct sockaddr_in *) rslv_ctx->addr.sockaddr;
+          *(p++) = sizeof(struct in_addr);
+          p = ngx_cpymem(p, &sin->sin_addr, sizeof(struct in_addr));
+    }
+
+not_found:
+
+    p = ngx_cpymem(p, rslv_ctx->name.data, rslv_ctx->name.len);
+    args.data = buf;
+    args.len = p - buf;
+
+    if (ngx_proxy_wasm_foreign_call_buffer_create(call, args.len) != NGX_OK) {
```

Comment:
When I look at the code of this `buffer_create` function now it seems to me like it should just be inlined. The core logic is just the buffer reuse, many of the lines in buffer_create can be avoided (e.g. the assertions), and this function is not reused anywhere else, so it feels like unnecessary redirection for the reader. I would say the same applies to `buffer_write` below as well.

Reply:
Yeah, well observed. I'll inline them.